keras_nlp/utils/pipeline_model.py

```diff
@@ -0,0 +1,284 @@
+        if y is not None:
+            y = self.preprocess_labels(y)
+        return pack_x_y_sample_weight(x, y, sample_weight)
+
+    # ========================================================================
+    # Below are overrides to keras.Model methods to apply the functions above.
+    # ========================================================================
+    def fit(
+        self,
+        x=None,
+        y=None,
+        batch_size=None,
+        sample_weight=None,
+        validation_data=None,
+        validation_split=None,
+        **kwargs,
+    ):
+        if validation_split and validation_data is None:
+            (x, y, sample_weight), validation_data = train_validation_split(
+                (x, y, sample_weight), validation_split=validation_split
+            )
+
+        x = convert_inputs_to_dataset(x, y, sample_weight, batch_size)
+        if self.include_preprocessing:
+            x = x.map(self.preprocess_samples, tf.data.AUTOTUNE)
+            x = x.prefetch(tf.data.AUTOTUNE)
+
+        if validation_data is not None:
+            if not isinstance(validation_data, tf.data.Dataset):
+                (vx, vy, vsw) = keras.utils.unpack_x_y_sample_weight(
+                    validation_data
+                )
+                validation_data = convert_inputs_to_dataset(
+                    vx, vy, vsw, batch_size
+                )
+
+        return super().fit(
+            x=x,
+            y=None,
+            batch_size=None,
+            sample_weight=None,
+            validation_data=validation_data,
+            **kwargs,
+        )
+
+    def evaluate(
+        self,
+        x=None,
+        y=None,
+        batch_size=None,
+        sample_weight=None,
+        **kwargs,
+    ):
+        # TODO: we can't support a cached eval dataset until we make changes to
+        # the upstream model. Otherwise we would cache the raw dataset.
+        kwargs.pop("_use_cached_eval_dataset", None)
+        x = convert_inputs_to_dataset(x, y, sample_weight, batch_size)
+        if self.include_preprocessing:
+            x = x.map(self.preprocess_samples, tf.data.AUTOTUNE)
+            x = x.prefetch(tf.data.AUTOTUNE)
+        return super().evaluate(
+            x=x,
+            y=None,
+            batch_size=None,
+            **kwargs,
+        )
+
+    def predict(
+        self,
+        x=None,
+        batch_size=None,
+        **kwargs,
+    ):
+        x = convert_inputs_to_dataset(x, None, None, batch_size)
+        if self.include_preprocessing:
+            x = x.map(self.preprocess_samples, tf.data.AUTOTUNE)
+            x = x.prefetch(tf.data.AUTOTUNE)
+
+        return super().predict(
+            x=x,
+            batch_size=None,
+            **kwargs,
+        )
+
+    def __call__(self, inputs, include_preprocessing=None, **kwargs):
+        # We don't trace if `including_preprocessing` is `False`, or we are
+        # currently tracing a functional model.
```

Comment:
What does "trace" mean in this context? I'm confused.

Reply:
Hmm, yeah maybe the wrong word. Wanted to avoid attempting to apply preprocessing when we are constructing a graph with functional model inputs, because applying preprocessing to functional inputs would never work.